### ISSUE TITLE
2.x

### ISSUE THREAD

**Issue by dereuromark** [joshuapaling/CakePHP-Export-CSV-Plugin#1]
For 2.x I recommend the https://github.com/josegonzalez/CsvView
Its cleaner to let the view scope handle the output formatting.
I guess your component was mainly a 1.x invention, right?

I wasn't able to comment in the bakery (503 error on post). Therefore the comment here.
You may close the issue, of course :)

**Comment by joshuapaling** [comment]
Hi Mark,

Thanks for your feedback. I hadn't seen Jose's plugin when I built mine. I just gave it a test run now.

The main difference is that mine handles nested belongsTo associations easily.

Example: Let's say you have `Suburb` belongsTo `Postcode`, which belongsTo `City`, which belongsTo `State`, which belongsTo `Country` (ie, 4 belongsTo's, nested). And say you wanted to export a CSV of all `Suburbs`, and all their associated info, right up to `Country`.

With my plugin, you simply use containable, do a `find('all')` and pass the plugin the resulting array. It will handle multiple nested belongsTo's no problem. See the recursively called `flattenArray` method in the source to see where that's handled. It will also handle cases where not all nested belongsTo associations are there for all rows (eg. one City record doesn't have a State specified, but all others do)

To download a CSV for the example I gave, the code would look like this:

```
$this->Suburb->contain(array(
    'Postcode' => array(
            'City' => array(
                    'State' => array(
                        'Country'
                    )
                )
        )
));

$suburbs = $this->Suburb->find('all');
$this->Export->exportCsv($suburbs, 'suburbs.csv');
```

So, despite it being a somewhat complex find call, there's only a single line of code specific to the plugin itself, which I think is nice.

With Jose's plugin, I think only the first level of belongsTo is handled easily out of the box. You'd have to do a custom view to handle Nested belongsTo's, and you'd have to do some recursive formatting like what I've done in my plugin (Edit: no, you don't - you can handle nested belongsTo's with the $_extract variable - still, it's a little more configuration than mine).

Basically, my plugin is good for exporting the results of a complex `find('all')` call, in a usable way, without having to explicitly specify how the CSV should be formatted. Pass it an array, and it just works everything out.

You're right, it should probably be handled in the view... I wrote it quickly, and didn't think to do it that way. At the moment, I kind of like the fact that the plugin requires only a single line to export a CSV. But, I'll consider moving it into the view if I can find the time.

On another note, being on the core team, do you know what the story is with the bakery website? It's super buggy, and there's no apparent way to sign up, change password, reset forgotten password, etc. Is anyone actively maintaining it? Is it open source? And open to pull requests?

**Comment by dereuromark** [comment]
Thank you for the detailed insight. That does indeed sound like a nice way to handle the data output.
Well, there seem to be always more than one solution :)

I am not involved with the bakery (maybe I should?). So I only know what you know - that it might need some more work to make it less buggy and more user friendly. But (like me) nobody probably found the time to do so yet. Especially with all the preparations for CakeFest and so on.

I think it is still running on 1.3 - https://github.com/cakephp/bakery - Feel free to help out.

**Comment by joshuapaling** [comment]
Update - I looked into Jose's plugin a bit more, and it turns out you can handle nested belongsTo's with the $_export variable. It's still more configuration than mine takes, but easier than I thought.

Probably what I should do is write a Component for Jose's plugin with a method to auto-prepare a nested array for the CSV View. When I find the time...

**Comment by joshuapaling** [comment]
Jose has just accepted my pull request, merging the best features of my plugin into his.
